source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc

```diff
@@ -169,16 +194,26 @@ Http::FilterDataStatus Filter::encodeData(Buffer::Instance& buffer, bool end_str
     auto& trailers = encoder_callbacks_->addEncodedTrailers();
     trailers.setGrpcStatus(grpc_status_);
 
-    if (withhold_grpc_frames_) {
+    if (withhold_grpc_frames_ && !response_size_header_) {
       buffer.prepend(buffer_);
-      buildGrpcFrameHeader(buffer);
+      buildGrpcFrameHeader(buffer, buffer.length());
     }
 
     return Http::FilterDataStatus::Continue;
   }
 
-  // We only need to buffer if we're responsible for injecting the gRPC frame header.
   if (withhold_grpc_frames_) {
+    if (response_size_header_) {
+      // If we're getting the response size from an upstream header, we can stream the response. The
+      // first chunk of data we encode needs the gRPC frame header prepended.
+      if (!frame_header_added_) {
+        buildGrpcFrameHeader(buffer, response_message_length_);
+        frame_header_added_ = true;
+      }
+
+      return Http::FilterDataStatus::Continue;
+    }
+
```

Comment:
Just thinking out loud: perhaps it would be simpler to call addEncodedData in encodeHeaders to inject the frame header  and then simply pass through all the data in encodeData? I'm not sure if this is a better approach (just alternate), so unless there is some real benefit to redoing it like that we can keep things as is